core/services/llo/channeldefinitions/onchain_channel_definition_cache.go

```diff
@@ -228,156 +358,299 @@ func (c *channelDefinitionCache) readLogs(ctx context.Context) (err error) {
+		c.lggr.Infow("Got new logs", "source", trigger.source, "url", trigger.url, "sha", hex.EncodeToString(trigger.sha[:]), "blockNum", trigger.blockNum)
+		c.fetchTriggerCh <- trigger
+	}
+}
 
-		err := channelConfigStoreABI.UnpackIntoInterface(unpacked, newChannelDefinitionEventName, log.Data)
-		if err != nil {
-			return fmt.Errorf("failed to unpack log data: %w", err)
-		}
-		if len(log.Topics) < 2 {
-			// should never happen but must guard against unexpected panics
-			c.lggr.Warnw("Log missing expected topics", "log", log)
-			continue
+// mergeDefinitions reconciles new channel definitions with the current set according to source
+// authority rules. Owner definitions (SourceOwner) have full authority: they can add, update, or
+// tombstone (delete) channels. Adder definitions (non-owner sources) have limited authority: they
+// can only add new channels and cannot overwrite or tombstone existing ones.
+//
+// Adder limits are enforced:
+//   - MaxChannelsPerAdder: The total number of channels in a single adder definition file cannot
+//     exceed this limit. This is checked before processing any channels.
+//   - MaxAdderAdditionsPerDefinition: The number of new channels an adder can add in a single
+//     definition is limited. Only channels that don't already exist are counted toward this limit.
+//     Existing channels from the same adder are skipped and not counted.
+//
+// Returns an error if adder limits are exceeded. The error will be one of:
+//   - errChannelsPerAdderLimitExceeded: When the definition file contains too many channels
+//   - errAdderAdditionsLimitExceeded: When the adder tries to add too many new channels
+func (c *channelDefinitionCache) mergeDefinitions(source uint32, currentDefinitions llotypes.ChannelDefinitions, newDefinitions llotypes.ChannelDefinitions) (llotypes.ChannelDefinitions, error) {
+	if source > SourceOwner {
+		if len(newDefinitions) > MaxChannelsPerAdder {
+			return nil, fmt.Errorf("%w: %d, max %d",
+				errChannelsPerAdderLimitExceeded, len(newDefinitions), MaxChannelsPerAdder)
 		}
-		unpacked.DonId = new(big.Int).SetBytes(log.Topics[1])
+	}
 
-		//nolint:gosec // disable G115
-		unpacked.Raw.BlockNumber = uint64(log.BlockNumber)
+	var adderAdditions int
+	for channelID, def := range newDefinitions {
+		switch def.Source {
+		case SourceOwner:
+			if def.Tombstone {
+				delete(currentDefinitions, channelID)
+				continue
+			}
+			currentDefinitions[channelID] = def
 
-		if unpacked.DonId.Cmp(big.NewInt(int64(c.donID))) != 0 {
-			// skip logs for other donIDs, shouldn't happen given the
-			// FilterLogs call, but belts and braces
-			continue
-		}
+		case source:
+			if def.Tombstone {
+				c.lggr.Warnw("invalid channel tombstone, cannot be added by adders",
+					"channelID", channelID, "adderID", def.Source)
+				continue
+			}
+			if existing, exists := currentDefinitions[channelID]; exists {
+				if existing.Source != def.Source {
```

Comment:
do we need this check here? wouldn't we want to emit the warning log anyways?

Reply:
The channel additions are not cumulative and each definition file represents the complete state of that source, we want to discard and log conflicts between sources, the existing definitions from an adder will always be present so we don't allow updates but it is normal for them to be always present.